src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAwaitForeachTests.cs

```diff
@@ -1944,18 +1942,14 @@ public async Task<bool> MoveNextAsync()
             i = i + 100; // Note: side-effects of async methods in structs are lost
             return more;
         }
-        public async ValueTask DisposeAsync()
-        {
-            Write($""Dispose({i}) "");
-            await Task.Yield();
-        }
+        public ValueTask DisposeAsync() => throw null;
     }
 }";
             var comp = CreateCompilationWithTasksExtensions(source + s_IAsyncEnumerable, options: TestOptions.DebugExe);
             comp.VerifyDiagnostics();
+            // Note: we only look for IAsyncDisposable, we don't look for a pattern-based DisposeAsync
```

Comment:
Is this by design? What for the `DisposeAsync` in `ConfiguredCancelableAsyncEnumerable<T>.Enumerator`, will it be ignored or not?

Reply:
> Is this by design?

Yes. That's a good observation though. 
@stephentoub, we need to make sure that the new types implementing `DisposeAsync` implement the `IAsyncDisposable` interface.

Reply:
I don't understand. How will ConfigureAwait work in that case?

Reply:
`ConfiguredCancelableAsyncEnumerable<T>.Enumerator` should implement `IAsyncDisposable`.
The `await foreach` binds `GetAsyncEnumerator`, `MoveNextAsync` and `Current` directly (pattern-based instead of via interface), but `DisposeAsync` must be implemented via interface.
The actual call to `DisposeAsync()` is a constrained call though, so no boxing.

Reply:
Oops, I had missed your point. I see the problem now.
`ConfiguredCancelableAsyncEnumerable<T>.Enumerator.DisposeAsync()` returns `ConfiguredValueTaskAwaitable`, which doesn't match the interface.

Filed tracking issue https://github.com/dotnet/roslyn/issues/32316